data/src/main/scala/cats/data/IList.scala

```diff
@@ -0,0 +1,276 @@
+    foldLeft(b)((acc, a) => B.combine(f(a), acc))
+
+  final def foldRight[B](b: B)(f: (A, B) => B): B =
+    reverse.foldLeft(b)((b, a) => f(a, b))
+
+  /** get the head if the [[IList]] is not empty */
+  final def headOption: Option[A] = this match {
+    case INil()      => Option.empty
+    case ICons(h, _) => Some(h)
+  }
+
+  /** check if an [[IList]] is empty */
+  final def isEmpty: Boolean = this match {
+    case INil()      => true
+    case ICons(_, _) => false
+  }
+
+  /** get the last element if the [[IList]] is not empty */
+  final def lastOption: Option[A] = {
+    @tailrec def loop(as: IList[A]): A = (as: @unchecked) match {
+      case ICons(h, INil()) => h
+      case ICons(_, t)      => loop(t)
+    }
+    if(isEmpty) None
+    else Some(loop(this))
+  }
+
+  /** get the element at the index if it exists */
+  final def lookup(index: Int): Option[A] = {
+    @tailrec
+    def loop(as: IList[A], i: Int): Option[A] = as match {
+      case ICons(h, t) =>
+        if(i > 0) loop(t, i - 1)
+        else if(i == 0) Some(h)
+        else None
+      case INil() => None
+    }
+    loop(this, index)
+  }
+
+  final def map[B](f: A => B): IList[B] =
+    reverse.foldLeft(empty[B])((acc, a) => ICons(f(a), acc))
+
+  /** add an element to the head of an [[IList]] */
+  final def prepend(a: A): IList[A] =
+    ICons(a, this)
+
+  /** alias for prepend */
+  final def ::(a: A): IList[A] =
+    ICons(a, this)
+
+  /** reverse an [[IList]] */
+  final def reverse: IList[A] =
+    foldLeft(empty[A])((acc, a) => ICons(a, acc))
+
+  /** provide a textual representation of an [[IList]] */
+  final def show(implicit A: Show[A]): String =
+    "[" + toNel.fold("")(_.reduceLeft(A.show)(_ + "," + A.show(_))) + "]"
+
+  /** compute the size of an [[IList]] */
+  final def size: Int = {
+    @inline @tailrec def loop(as: IList[A], acc: Int): Int = as match {
+      case ICons(_, t) => loop(t, acc + 1)
+      case INil()      => acc
+    }
+    loop(this, 0)
+  }
+
+  /** get the tail if the [[IList]] is not empty */
+  final def tailOption: Option[IList[A]] = this match {
+    case INil()      => Option.empty
+    case ICons(_, t) => Some(t)
+  }
+
+  /** take the `n` first elements */
+  final def take(n: Int): IList[A] = {
+    @tailrec
+    def loop(as: IList[A], m: Int, acc: IList[A]): IList[A] = as match {
+      case INil() => acc.reverse
+      case ICons(h,t) => if(m > 0) loop(t, m - 1, ICons(h, acc)) else acc.reverse
+    }
+    loop(this, n, empty)
+  }
+
+  /** take elements as long as the predicate holds */
+  final def takeWhile(p: A => Boolean): IList[A] = {
+    @tailrec
+    def loop(as: IList[A], acc: IList[A]): IList[A] = as match {
+      case INil() => acc.reverse
+      case ICons(h,t) => if(p(h)) loop(t, ICons(h, acc)) else acc.reverse
+    }
+    loop(this, empty)
+  }
+
+  /** transform an [[IList]] into a [[scala.List]] */
+  final def toList: List[A] =
+    reverse.foldLeft(List.empty[A])((acc, a) => a :: acc)
```

Comment:
You could speed this up by using `foldLeft(ListBuffer.empty[A])(...)` instead of having to reverse the `IList`.